src/shared/orgListUtil.ts

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { basename, join } from 'path';
+
+import { Org, AuthInfo, fs, sfdc, ConfigAggregator, Global, AuthFields, Logger } from '@salesforce/core';
+import { Dictionary, JsonMap } from '@salesforce/ts-types';
+import { Record } from 'jsforce';
+import { omit } from '@salesforce/kit/lib';
+import { getAliasByUsername } from './utils';
+import { ScratchOrgInfoSObject, ExtendedAuthFields } from './orgTypes';
+
+type OrgGroups = {
+  nonScratchOrgs: ExtendedAuthFields[];
+  scratchOrgs: ExtendedAuthFields[];
+};
+
+type ExtendedScratchOrgInfo = ScratchOrgInfoSObject & {
+  devHubOrgId: string;
+};
+
+export class OrgListUtil {
+  private static logger;
+
+  public static async retrieveLogger(): Promise<Logger> {
+    if (!OrgListUtil.logger) {
+      OrgListUtil.logger = await Logger.child('OrgListUtil');
+    }
+    return OrgListUtil.logger;
+  }
+
+  /**
+   * This method takes all locally configured orgs and organizes them into the following buckets:
+   * { nonScratchOrgs: [{}], scratchOrgs: [{}] }
+   * the scratchOrgInfo query.
+   *
+   * @param {string[]|null} userFilenames- an array of strings that are validated against the server.
+   * @param {object} flags - the result of this.flags on an sfdx command
+   */
+  public static async readLocallyValidatedMetaConfigsGroupedByOrgType(
+    userFilenames: string[],
+    flags
+  ): Promise<OrgGroups> {
+    const contents: AuthInfo[] = await this.readAuthFiles(userFilenames);
+    const orgs = await this.groupOrgs(contents);
+
+    // parallelize two very independent operations
+    const [nonScratchOrgs, scratchOrgs] = await Promise.all([
+      Promise.all(
+        orgs.nonScratchOrgs.map(async (fields) => {
+          if (flags.skipconnectionstatus) {
+            fields.connectedStatus = fields.connectedStatus ?? 'Unknown';
+          } else {
+            fields.connectedStatus = await this.determineConnectedStatusForNonScratchOrg(fields.username);
+          }
+          return fields;
+        })
+      ),
+
+      this.processScratchOrgs(orgs.scratchOrgs),
+    ]);
+
+    return {
+      nonScratchOrgs,
+      scratchOrgs,
+    };
+  }
+
+  public static async processScratchOrgs(scratchOrgs: ExtendedAuthFields[]): Promise<ExtendedAuthFields[]> {
+    // organize by DevHub to reduce queries
+    const orgIdsGroupedByDevHub: Dictionary<string[]> = {};
+    scratchOrgs.forEach((fields) => {
```

Comment:
say more about what you have in mind?  This is a kinda like groupby but instead of grouping the whole object, grab one property and do a small transform (18=>15 character IDs)

Reply:
Sorry for not being more clear. Thinking of a lambda that accomplishes the same thing as this logic, produce a dictionary by devhub username and its associated list of scratch org ids. JS reduce does not neatly handle this as well as Java Stream, so the result is not much different, other than eliminating conditionals and removes initialization of the result var.

My example for this
```
    const orgIdsGroupedByDevHub = scratchOrgs
      .filter((fields) => fields.devHubUsername)
      .reduce((accumulator: Dictionary<string[]>, fields) => {
        const orgs = accumulator[fields.devHubUsername] ?? [];
        orgs.push(sfdc.trimTo15(fields.orgId));
        accumulator[fields.devHubUsername] = orgs;
        return accumulator;
      }, {});
```

Reply:
I changed it to something more along those lines.